src/main/java/com/google/api/generator/engine/ast/TypeNode.java

```diff
@@ -124,9 +146,10 @@ public boolean equals(Object o) {
     }
 
     TypeNode type = (TypeNode) o;
-    return typeKind().equals(type.typeKind())
-        && (isArray() == type.isArray())
-        && Objects.equals(reference(), type.reference());
+    return isBoxedTypeEquals(type, this)
```

Comment:
Could we please add parentheses around logic groups for clarity? Also, what if one of the boxed types are arrays (which needs a test as well)?

Reply:
of course, good point. Wrapped the strict equality check to `strictEquals()` method. 
`int[]` cannot be cast to `Integer[]` so boxed type array is not equal to its primitive type array. This check is included in `isBoxedTypeEquals()` method, since we the TypeNode we put in the map is without array setting. Added a unit test to verify that. Thanks